File: ios/calorietracker/Models/IngredientPortion.swift
import Foundation

/// Unrounded nutrition baseline used while typing a new ingredient weight.
struct IngredientPortion: Equatable, Sendable {
    var grams: Double
    var calories: Double
    var protein: Double
    var carbs: Double
    var fat: Double

    nonisolated init(grams: Double, calories: Double, protein: Double, carbs: Double, fat: Double) {
        self.grams = grams
        self.calories = calories
        self.protein = protein
        self.carbs = carbs
        self.fat = fat
    }

    nonisolated func resized(to newGrams: Double) -> IngredientPortion? {
        guard grams.isFinite, grams > 0, newGrams.isFinite, newGrams > 0 else { return nil }
        let factor = newGrams / grams
        let result = IngredientPortion(grams: newGrams, calories: calories * factor,
                                      protein: protein * factor, carbs: carbs * factor, fat: fat * factor)
        guard [result.calories, result.protein, result.carbs, result.fat].allSatisfy({ $0.isFinite && $0 >= 0 && $0 < Double(Int32.max) }) else { return nil }
        return result
    }
}

File: android/app/src/main/java/com/ayuvo/health/models/IngredientPortion.kt
package com.ayuvo.health.models

/** Unrounded nutrition baseline used while typing a new ingredient weight. */
data class IngredientPortion(
    val grams: Double,
    val calories: Double,
    val protein: Double,
    val carbs: Double,
    val fat: Double
) {
    fun resized(newGrams: Double): IngredientPortion? {
        if (!grams.isFinite() || grams <= 0 || !newGrams.isFinite() || newGrams <= 0) return null
        val factor = newGrams / grams
        val result = IngredientPortion(newGrams, calories * factor, protein * factor, carbs * factor, fat * factor)
        if (listOf(result.calories, result.protein, result.carbs, result.fat).any { !it.isFinite() || it < 0 || it >= Int.MAX_VALUE.toDouble() }) return null
        return result
    }
}
